File: src/c++/dnp3/outstation.cpp
#include <iostream>

#include "outstation.hpp"

#include "fmt/format.h"
#include "opendnp3/outstation/UpdateBuilder.h"

namespace otsim {
namespace dnp3 {

Outstation::Outstation(OutstationConfig config, OutstationRestartConfig restart, Pusher pusher) :
  DefaultOutstationApplication(opendnp3::TimeDuration::Minutes(1)), config(config), restartConfig(restart), pusher(pusher)
{
  metrics = otsim::msgbus::MetricsPusher::Create();
  metrics->NewMetric("Counter", "status_count", "number of status messages processed");
}

opendnp3::OutstationStackConfig Outstation::Init() {
  opendnp3::DatabaseConfig db = opendnp3::DatabaseConfig();

  for (const auto& kv : binaryPoints) {
    db.binary_input[kv.first] = {};

    db.binary_input[kv.first].svariation = kv.second.svariation;
    db.binary_input[kv.first].evariation = kv.second.evariation;
    db.binary_input[kv.first].clazz = kv.second.clazz;
  }

  for (const auto& kv : analogPoints) {
    db.analog_input[kv.first] = {};

    db.analog_input[kv.first].svariation = kv.second.svariation;
    db.analog_input[kv.first].evariation = kv.second.evariation;
    db.analog_input[kv.first].clazz = kv.second.clazz;
    db.analog_input[kv.first].deadband = kv.second.deadband;
  }

  opendnp3::OutstationStackConfig stack(db);

  stack.outstation.eventBufferConfig = opendnp3::EventBufferConfig::AllTypes(100);
  stack.link.LocalAddr  = config.localAddr;
  stack.link.RemoteAddr = config.remoteAddr;

  return stack;
}

void Outstation::Run() {
  metrics->Start(pusher, config.id);

  while (running) {
    if (restartConfig.coldRestart) {
      restartConfig.coldRestarter(config.localAddr);
      restartConfig.coldRestart.store(false);

      continue;
    }

    if (restartConfig.warmRestart) {
      Disable();
      std::this_thread::sleep_for(std::chrono::seconds(restartConfig.warm));
      Enable();

      restartConfig.warmRestart.store(false);
      continue;
    }

    opendnp3::UpdateBuilder builder;

    for (const auto& kv : binaryPoints) {
      const std::uint16_t addr = kv.first;
      const std::string   tag  = kv.second.tag;

      try {
        auto lock = std::unique_lock<std::mutex>(pointsMu);

        auto point = points.at(tag);
        builder.Update(opendnp3::Binary(point.value != 0), addr);

        std::cout << fmt::format("[{}] updated binary {} to {}", config.id, addr, point.value) << std::endl;
      } catch (const std::out_of_range&) {}
    }

    for (const auto& kv : analogPoints) {
      const std::uint16_t addr = kv.first;
      const std::string   tag  = kv.second.tag;

      try {
        auto lock = std::unique_lock<std::mutex>(pointsMu);

        auto point = points.at(tag);
        builder.Update(opendnp3::Analog(point.value, opendnp3::Flags(0), opendnp3::DNPTime(point.ts)), addr);

        std::cout << fmt::format("[{}] updated analog {} to {}", config.id, addr, point.value) << std::endl;
      } catch (const std::out_of_range&) {}
    }

    outstation->Apply(builder.Build());
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }

  metrics->Stop();
}

bool Outstation::AddBinaryInput(BinaryPoint point) {
  binaryPoints[point.address] = point;
  points[point.tag] = otsim::msgbus::Point{point.tag, 0.0, 0};

  return true;
}

bool Outstation::AddBinaryOutput(BinaryPoint point) {
  point.output = true;

  binaryPoints[point.address] = point;
  points[point.tag] = otsim::msgbus::Point{point.tag, 0.0, 0};

  return true;
}

bool Outstation::AddAnalogInput(AnalogPoint point) {
  analogPoints[point.address] = point;
  points[point.tag] = otsim::msgbus::Point{point.tag, 0.0, 0};

  return true;
}

bool Outstation::AddAnalogOutput(AnalogPoint point) {
  point.output = true;

  analogPoints[point.address] = point;
  points[point.tag] = otsim::msgbus::Point{point.tag, 0.0, 0};

  return true;
}

void Outstation::WriteBinary(std::uint16_t address, bool status) {
  auto iter = binaryPoints.find(address);
  if (iter == binaryPoints.end()) {
    return;
  }

  std::cout << fmt::format("[{}] setting tag {} to {}", config.id, iter->second.tag, status) << std::endl;

  otsim::msgbus::Points points;
  points.push_back(otsim::msgbus::Point{iter->second.tag, status ? 1.0 : 0.0});

  otsim::msgbus::Update contents = {.updates = points};
  auto env = otsim::msgbus::NewEnvelope(config.id, contents);

  pusher->Push("RUNTIME", env);
}

void Outstation::WriteAnalog(std::uint16_t address, double value) {
  auto iter = analogPoints.find(address);
  if (iter == analogPoints.end()) {
    return;
  }

  std::cout << fmt::format("[{}] setting tag {} to {}", config.id, iter->second.tag, value) << std::endl;

  otsim::msgbus::Points points;
  points.push_back(otsim::msgbus::Point{iter->second.tag, value});

  otsim::msgbus::Update contents = {.updates = points};
  auto env = otsim::msgbus::NewEnvelope(config.id, contents);

  pusher->Push("RUNTIME", env);
}

const BinaryPoint* Outstation::GetBinary(const uint16_t address) {
  auto iter = binaryPoints.find(address);
  if (iter == binaryPoints.end()) {
    return NULL;
  }

  return &iter->second;
}

const AnalogPoint* Outstation::GetAnalog(const uint16_t address) {
  auto iter = analogPoints.find(address);
  if (iter == analogPoints.end()) {
    return NULL;
  }

  return &iter->second;
}

void Outstation::ResetOutputs() {
  otsim::msgbus::Points points;

  for (const auto& kv : binaryPoints) {
    if (kv.second.output) {
      points.push_back(otsim::msgbus::Point{kv.second.tag, 0.0});
    }
  }

  for (const auto& kv : analogPoints) {
    if (kv.second.output) {
      points.push_back(otsim::msgbus::Point{kv.second.tag, 0.0});
    }
  }

  if (points.size()) {
    std::cout << fmt::format("[{}] setting outputs to zero values", config.id) << std::endl;

    otsim::msgbus::Update contents = {.updates = points};
    auto env = otsim::msgbus::NewEnvelope(config.id, contents);

    pusher->Push("RUNTIME", env);
  }
}

void Outstation::HandleMsgBusStatus(const otsim::msgbus::Envelope<otsim::msgbus::Status>& env) {
  auto sender = otsim::msgbus::GetEnvelopeSender(env);

  if (sender == config.id) {
    return;
  }

  metrics->IncrMetric("status_count");

  for (auto &p : env.contents.measurements) {
    if (points.count(p.tag)) {
      std::cout << fmt::format("[{}] status received for tag {}", config.id, p.tag) << std::endl;

      auto lock = std::unique_lock<std::mutex>(pointsMu);
      points[p.tag] = p;
    }
  }
}

uint16_t Outstation::ColdRestart() {
  restartConfig.coldRestart.store(true);
  return restartConfig.cold;
}

uint16_t Outstation::WarmRestart() {
  restartConfig.warmRestart.store(true);
  return restartConfig.warm;
}

opendnp3::CommandStatus Outstation::Select(const opendnp3::ControlRelayOutputBlock& arCommand, std::uint16_t aIndex) {
    if (!GetBinary(aIndex)) {
        // This is our best guess at what status to return when the address
        // being selected doesn't exist locally.
        return opendnp3::CommandStatus::OUT_OF_RANGE;
    }

    return opendnp3::CommandStatus::SUCCESS;
}

opendnp3::CommandStatus Outstation::Operate(const opendnp3::ControlRelayOutputBlock& arCommand, std::uint16_t aIndex, opendnp3::IUpdateHandler& handler, opendnp3::OperateType opType) {
    auto point = GetBinary(aIndex);

    if (!point) {
        // This is our best guess at what status to return when the address
        // being selected doesn't exist locally.
        return opendnp3::CommandStatus::OUT_OF_RANGE;
    }

    if (point->sbo && opType != opendnp3::OperateType::SelectBeforeOperate) {
        return opendnp3::CommandStatus::NO_SELECT;
    }

    bool val;

    switch (arCommand.opType) {
      case opendnp3::OperationType::LATCH_ON:
        val = true;
        break;
      case opendnp3::OperationType::LATCH_OFF:
        val = false;
        break;
      case opendnp3::OperationType::PULSE_ON:
        switch(arCommand.tcc) {
          case opendnp3::TripCloseCode::TRIP:
            val = false;
            break;
          case opendnp3::TripCloseCode::CLOSE:
            val = true;
            break;
          default:
            return opendnp3::CommandStatus::NOT_SUPPORTED;
        }

        break;
      default:
        return opendnp3::CommandStatus::NOT_SUPPORTED;
    }

    WriteBinary(aIndex, val);
    return opendnp3::CommandStatus::SUCCESS;
}

opendnp3::CommandStatus Outstation::Select(const opendnp3::AnalogOutputFloat32& arCommand, std::uint16_t aIndex) {
    if (!GetAnalog(aIndex)) {
        // This is our best guess at what status to return when the address
        // being selected doesn't exist locally.
        return opendnp3::CommandStatus::OUT_OF_RANGE;
    }

    return opendnp3::CommandStatus::SUCCESS;
}

opendnp3::CommandStatus Outstation::Operate(const opendnp3::AnalogOutputFloat32& arCommand, std::uint16_t aIndex, opendnp3::IUpdateHandler& handler, opendnp3::OperateType opType) {
    auto point = GetAnalog(aIndex);

    if (!point) {
        // This is our best guess at what status to return when the address
        // being selected doesn't exist locally.
        return opendnp3::CommandStatus::OUT_OF_RANGE;
    }

    if (point->sbo && opType != opendnp3::OperateType::SelectBeforeOperate) {
        return opendnp3::CommandStatus::NO_SELECT;
    }

    WriteAnalog(aIndex, arCommand.value);
    return opendnp3::CommandStatus::SUCCESS;
}

} // namespace dnp3
} // namespace otsim

File: src/old/python/dnp3/outstation.py
from __future__  import annotations

import json, time, typing, zmq

from collections import namedtuple
from pydnp3      import asiodnp3, opendnp3, openpal

import envelope
from point import Point

class OutstationConfig(typing.NamedTuple):
    id: str
    local_addr: int
    remote_addr: int
    cold_restart_delay: int
    warm_restart_delay: int
    cold_restarter: ColdRestartFunc

ColdRestartFunc = typing.Callable[[int], None]

class Outstation(opendnp3.IOutstationApplication, opendnp3.ICommandHandler):
  def __init__(self: Outstation, config: OutstationConfig, pusher: zmq.Socket):
    opendnp3.IOutstationApplication.__init__(self)
    opendnp3.ICommandHandler.__init__(self)

    self.config = config
    self.pusher = pusher

    self.cold_restart: bool = False
    self.warm_restart: bool = False

    # will be set by the server
    self.iOutstation: opendnp3.IOutstation = None

    self.binary_inputs:  typing.List[Point] = []
    self.binary_outputs: typing.List[Point] = []
    self.analog_inputs:  typing.List[Point] = []
    self.analog_outputs: typing.List[Point] = []

  def init_stack_config(self: Outstation) -> asiodnp3.OutstationStackConfig:
    sizes = opendnp3.DatabaseSizes(
      len(self.binary_inputs),
      0,
      len(self.analog_inputs),
      0,
      0,
      len(self.binary_outputs),
      len(self.analog_outputs),
      0,
    )

    stack_config = asiodnp3.OutstationStackConfig(sizes)

    stack_config.outstation.eventBufferConfig = opendnp3.EventBufferConfig().AllTypes(100)
    stack_config.outstation.params.allowUnsolicited = True
    stack_config.link.LocalAddr = self.config.local_addr
    stack_config.link.RemoteAddr = self.config.remote_addr
    stack_config.link.KeepAliveTimeout = openpal.TimeDuration().Max()

    self.init_db(stack_config.dbConfig)

    return stack_config

  def init_db(self: Outstation, db_config: asiodnp3.DatabaseConfig) -> None:
    # TODO: I don't think this is actually configuring the database entries for some reason...

    for idx, p in enumerate(self.binary_inputs):
      db_config.binary[idx].vIndex = p.address
      db_config.binary[idx].svariation = p.svariation
      db_config.binary[idx].evariation = p.evariation
      db_config.binary[idx].clazz = p.clazz

    for idx, p in enumerate(self.analog_inputs):
      db_config.analog[idx].vIndex = p.address
      db_config.analog[idx].svariation = p.svariation
      db_config.analog[idx].evariation = p.evariation
      db_config.analog[idx].clazz = p.clazz
      db_config.analog[idx].deadband = p.deadband

    for idx, p in enumerate(self.binary_outputs):
      db_config.boStatus[idx].vIndex = p.address
      db_config.boStatus[idx].svariation = p.svariation
      db_config.boStatus[idx].evariation = p.evariation
      db_config.boStatus[idx].clazz = p.clazz

    for idx, p in enumerate(self.analog_outputs):
      db_config.aoStatus[idx].vIndex = p.address
      db_config.aoStatus[idx].svariation = p.svariation
      db_config.aoStatus[idx].evariation = p.evariation
      db_config.aoStatus[idx].clazz = p.clazz
      db_config.aoStatus[idx].deadband = p.deadband

  def run(self: Outstation, tags: typing.Dict[str, float]) -> None:
    while True:
      # TODO: what happens in this loop when this outstation isn't the one that
      # initiated the cold restart? Will the disabling of the DNP3 IOutstation
      # by the server be enough?
      if self.cold_restart:
        # this will block for duration of cold restart
        self.config.cold_restarter(self.config.local_addr)
        self.cold_restart = False
        continue

      if self.warm_restart:
        self.disable()
        time.sleep(self.config.warm_restart_delay)
        self.enable()

        self.warm_restart = False
        continue

      builder = asiodnp3.UpdateBuilder()

      for idx, point in enumerate(self.binary_inputs):
        status = tags.get(point.tag, 0)
        # TODO: figure out best way to track time for points
        builder.Update(opendnp3.Binary(status != 0), idx)

      for idx, point in enumerate(self.binary_outputs):
        status = tags.get(point.tag, 0)
        # TODO: figure out best way to track time for points
        builder.Update(opendnp3.BinaryOutputStatus(status != 0), idx)

      for idx, point in enumerate(self.analog_inputs):
        value = tags.get(point.tag, 0.0)
        # TODO: figure out best way to track time for points
        builder.Update(opendnp3.Analog(value, opendnp3.Flags(0), opendnp3.DNPTime(0)), idx)

      for idx, point in enumerate(self.analog_outputs):
        value = tags.get(point.tag, 0.0)
        # TODO: figure out best way to track time for points
        builder.Update(opendnp3.AnalogOutputStatus(value, opendnp3.Flags(0), opendnp3.DNPTime(0)), idx)

      update = builder.Build()
      self.iOutstation.Apply(update)

      time.sleep(1)

  def add_binary(self: Outstation, point: Point) -> bool:
    if point.output:
      self.binary_outputs.append(point)
    else:
      self.binary_inputs.append(point)

    return True

  def add_analog(self: Outstation, point: Point) -> bool:
    if point.output:
      self.analog_outputs.append(point)
    else:
      self.analog_inputs.append(point)

    return True

  def reset_outputs(self: Outstation) -> None:
    for point in self.binary_outputs:
      update: envelope.Point = {'tag': point.tag, 'value': 0.0}
      env = envelope.new_update_envelope('dnp3', {'updates': [update]})

      self.pusher.send_multipart((b'RUNTIME', json.dumps(env).encode()))

    for point in self.analog_outputs:
      update: envelope.Point = {'tag': point.tag, 'value': 0.0}
      env = envelope.new_update_envelope('dnp3', {'updates': [update]})

      self.pusher.send_multipart((b'RUNTIME', json.dumps(env).encode()))

  def enable(self: Outstation) -> bool:
    return self.iOutstation.Enable()

  def disable(self: Outstation) -> bool:
    return self.iOutstation.Disable()

  # Overridden IOutstationApplication method
  def ColdRestartSupport(self: Outstation) -> opendnp3.RestartMode:
    return opendnp3.RestartMode.SUPPORTED_DELAY_COARSE

  # Overridden IOutstationApplication method
  def ColdRestart(self: Outstation) -> int:
    self.cold_restart = True
    return self.config.cold_restart_delay

  # Overridden IOutstationApplication method
  def WarmRestartSupport(self: Outstation) -> opendnp3.RestartMode:
    return opendnp3.RestartMode.SUPPORTED_DELAY_COARSE

  # Overridden IOutstationApplication method
  def WarmRestart(self: Outstation) -> int:
    self.warm_restart = True
    return self.config.warm_restart_delay

  # Overridden ICommandHandler method
  def Start(self: Outstation) -> None:
    pass

  # Overridden ICommandHandler method
  def End(self: Outstation) -> None:
    pass

  # Overridden ICommandHandler method
  def Select(self: Outstation, command: typing.Any, index: int) -> opendnp3.CommandStatus:
    if isinstance(command, opendnp3.ControlRelayOutputBlock):
      point = next(filter(lambda p: p.address == index, self.binary_outputs), None)

      if not point:
        return opendnp3.CommandStatus.OUT_OF_RANGE

      return opendnp3.CommandStatus.SUCCESS

    if isinstance(command, opendnp3.AnalogOutputFloat32):
      point = next(filter(lambda p: p.address == index, self.analog_outputs), None)

      if not point:
        return opendnp3.CommandStatus.OUT_OF_RANGE

      return opendnp3.CommandStatus.SUCCESS

    return opendnp3.CommandStatus.NOT_SUPPORTED

  # Overridden ICommandHandler method
  def Operate(self: Outstation, command: typing.Any, index: int, op_type: opendnp3.OperateType) -> opendnp3.CommandStatus:
    if isinstance(command, opendnp3.ControlRelayOutputBlock):
      point = next(filter(lambda p: p.address == index, self.binary_outputs), None)

      if not point:
        return opendnp3.CommandStatus.OUT_OF_RANGE

      if point.sbo and op_type is not opendnp3.OperateType.SELECT_BEFORE_OPERATE:
        return opendnp3.CommandStatus.NO_SELECT

      status: float = 1.0 if command.functionCode == opendnp3.ControlCode.LATCH_ON else 0.0

      update: envelope.Point = {'tag': point.tag, 'value': status}
      env = envelope.new_update_envelope('dnp3', {'updates': [update]})

      self.pusher.send_multipart((b'RUNTIME', json.dumps(env).encode()))

      return opendnp3.CommandStatus.SUCCESS

    if isinstance(command, opendnp3.AnalogOutputFloat32):
      point = next(filter(lambda p: p.address == index, self.binary_outputs), None)

      if not point:
        return opendnp3.CommandStatus.OUT_OF_RANGE

      if point.sbo and op_type is not opendnp3.OperateType.SELECT_BEFORE_OPERATE:
        return opendnp3.CommandStatus.NO_SELECT

      value: float = command.value

      update: envelope.Point = {'tag': point.tag, 'value': value}
      env = envelope.new_update_envelope('dnp3', {'updates': [update]})

      self.pusher.send_multipart((b'RUNTIME', json.dumps(env).encode()))

      return opendnp3.CommandStatus.SUCCESS

    return opendnp3.CommandStatus.NOT_SUPPORTED
